sdk/nodejs/policy/proxy.ts

```diff
@@ -0,0 +1,106 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export function proxy<T>(toProxy: any): any {
+    return proxyHelper(toProxy, []);
+}
+
+function proxyHelper<T>(toProxy: any, props: (keyof T)[]): any {
+    if (!(toProxy instanceof Object)) {
+        return toProxy;
+    }
+
+    return new Proxy(toProxy, {
+        get: (obj, prop: keyof T) => {
+            const newProps = props.concat([prop]);
+            const field = obj[prop];
+            if (isUnknown(field)) {
+                throw new UnknownValueError<T>(field, newProps);
+            }
+            return proxyHelper(field, newProps);
+        },
+    });
+}
+
+export class UnknownValueError<T> extends Error {
+    constructor(public readonly unknownType: string, public readonly props: (keyof T)[]) {
+        super(
+            `${unknownToString(unknownType)} value at .${props.join(
+                ".",
+            )} can't be statically computed`,
```

Comment:
nit: just make this a one-liner :)

Reply:
I defer all style issues to the formatter—if this is _really_ important I can maybe shorten one of the variables.

Reply:
Let's pull some of those computations out into local variables.